subsys/fb/cfb_shell.c

```diff
@@ -0,0 +1,355 @@
+/** @file
+ * @brief Monochrome Character Framebuffer shell module
+ *
+ * Provide some Character Framebuffer shell commands that can be useful for
+ * testing.
+ */
+
+/*
+ * Copyright (c) 2018 Diego Sueiro
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdlib.h>
+#include <shell/shell.h>
+#include <display/cfb.h>
+
+#define HELP_NONE "[none]"
+#define HELP_INIT "call \"cfb init\" first"
+
+#define DISPLAY_DRIVER	CONFIG_CHARACTER_FRAMEBUFFER_SHELL_DRIVER_NAME
+
+
+static struct device *dev;
+
+static int cmd_clear(const struct shell *shell, size_t argc, char *argv[])
+{
+	int err;
+
+	if (!dev) {
+		shell_error(shell, HELP_INIT);
+		return -ENODEV;
+	}
+
+	err = cfb_framebuffer_clear(dev, true);
+	if (err) {
+		shell_error(shell, "Framebuffer clear error=%d", err);
+		return err;
+	}
+
+	err = cfb_framebuffer_finalize(dev);
+	if (err) {
+		shell_error(shell, "Framebuffer finalize error=%d", err);
+		return err;
+	}
+
+	shell_print(shell, "Display Cleared");
+
+	return err;
+}
+
+static int cmd_cfb_print(const struct shell *shell, int col, int row, char *str)
+{
+	int err;
+	u8_t ppt;
+
+	if (!dev) {
+		shell_error(shell, HELP_INIT);
+		return -ENODEV;
+	}
+
+	ppt = cfb_get_display_parameter(dev, CFB_DISPLAY_PPT);
+
+	err = cfb_framebuffer_clear(dev, false);
+	if (err) {
+		shell_error(shell, "Framebuffer clear failed error=%d", err);
+		return err;
+	}
+
+	err = cfb_print(dev, str, col, row * ppt);
+	if (err) {
+		shell_error(shell, "Failed to print the string %s error=%d",
+		      str, err);
+		return err;
+	}
+
+	err = cfb_framebuffer_finalize(dev);
+	if (err) {
+		shell_error(shell,
+			    "Failed to finalize the Framebuffer error=%d", err);
+		return err;
+	}
+
+	return err;
+}
+
+static int cmd_print(const struct shell *shell, size_t argc, char *argv[])
+{
+	int err;
+	int col, row;
+
+	if (!dev) {
+		shell_error(shell, HELP_INIT);
+		return -ENODEV;
+	}
+
+	col = strtol(argv[1], NULL, 10);
+	if (col > cfb_get_display_parameter(dev, CFB_DISPLAY_COLS)) {
+		shell_error(shell, "Invalid col=%d position", col);
+		return -EINVAL;
+	}
+
+	row = strtol(argv[2], NULL, 10);
+	if (row > cfb_get_display_parameter(dev, CFB_DISPLAY_ROWS)) {
+		shell_error(shell, "Invalid row=%d position", row);
+		return -EINVAL;
+	}
+
+	err = cmd_cfb_print(shell, col, row, argv[3]);
+	if (err) {
+		shell_error(shell, "Failed printing to Framebuffer error=%d",
+			    err);
+	}
+
+	return err;
+}
+
+static int cmd_scroll(const struct shell *shell, size_t argc, char *argv[])
+{
+	int err = 0;
+	int col, row;
+	int scroll, boundary;
+
+	if (!dev) {
+		shell_error(shell, HELP_INIT);
+		return -ENODEV;
+	}
+
+	if (!strcmp(argv[1], "vertical")) {
```

Comment:
It is not bad implementation and you can keep it as it is.

From my experience I would use a different convention. 
I would create a subcommands `vertical` and `horizontal` and I would implement handlers there. It will simplify usage from end user perspective. User will be able to prompt / autocomplete them. In this scenario handler `cmd_scroll` would be not needed at all.

Another solution / compromise would be adding subcommands `vertical` and `horizontal` without a handler. User will be able to autocomplete them with the `tab` key but handler `cmd_scroll` could stay as it is.

Reply:
Can you please point me some code examples using the subcommands?

Reply:
Well, you are using subcommands in this example, you create them with macro:  `SHELL_CREATE_STATIC_SUBCMD_SET` :)

You can change you have defined:

`SHELL_CMD_ARG(scroll, NULL, "<dir: (vertical|horizontal)> " "<col: pos> <row: pos> <text>", cmd_scroll, 5, 0),`

you can change to (I've added **&sub_scroll**):

`SHELL_CMD_ARG(scroll, &sub_scroll, "<dir: (vertical|horizontal)> " "<col: pos> <row: pos> <text>", cmd_scroll, 5, 0),`

Next you need to create  **&sub_scroll** like that:

```
SHELL_CREATE_STATIC_SUBCMD_SET(sub_scroll) {
    SHELL_CMD_ARG(horizontal, NULL, HELP_NONE, NULL),
    SHELL_CMD_ARG(vertical, NULL, HELP_NONE, NULL),
    SHELL_SUBCMD_SET_END
};
```

It is important that `SHELL_CREATE_STATIC_SUBCMD_SET(sub_scroll)` is defined **above** : `SHELL_CMD_ARG(scroll, &sub_scroll, "<dir: (vertical|horizontal)> " "<col: pos> <row: pos> <text>", cmd_scroll, 5, 0),`

**EXAMPLE:**
You can find in file `subsys/shell/shell_cmds.c`
Take a look at command `shell`. You can call it `shell colors off`:

- shell - root command
- colors - subcommand of shell command
- off  - subcommand of colors command

Reply:
Understood.
But when calling the `cmd_scroll` I still need to know if it is `vertical` or `horizontal` for example.
If I don't have a handler for each subcommand and use the `cmd_scroll` I still need to use the `strcmp`. Right?

Maybe instead of having a  `cmd_scroll` I add a  `sub_cmd_scroll_vert` and  `sub_cmd_scroll_horz`.

What do you say?

Reply:
Yes, if you wont implement handlers for `vertical` and `horizontal` you will still need to use `strcmp`. However you will be able to complete / prompt them.

I am usually implementing handlers (like for colors subcommands: `on`, and `off')

Reply:
> Maybe instead of having a cmd_scroll I add a sub_cmd_scroll_vert and sub_cmd_scroll_horz.

This is great idea that I support ;)